datasource.go

```diff
@@ -0,0 +1,204 @@
+package main
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"strconv"
+	"strings"
+
+	"github.com/bitly/go-simplejson"
+	"github.com/grafana/grafana-plugin-model/go/datasource"
+	plugin "github.com/hashicorp/go-plugin"
+	"golang.org/x/net/context"
+	"golang.org/x/net/context/ctxhttp"
+)
+
+var httpClient = &http.Client{}
+
+type ClickhouseDatasource struct {
+	plugin.NetRPCUnsupportedPlugin
+}
+
+func (t *ClickhouseDatasource) Query(ctx context.Context, req *datasource.DatasourceRequest) (r *datasource.DatasourceResponse, err error) {
+	// catch all panics and override err return value
+	defer func() {
+		if r := recover(); r != nil {
+			err = fmt.Errorf("clickhouse plugin panicked: %#w", r)
+		}
+	}()
+
+	refId := req.Queries[0].RefId
+	modelJson, err := simplejson.NewJson([]byte(req.Queries[0].ModelJson))
+	if err != nil {
+		return nil, err
+	}
+
+	query := modelJson.Get("rawQuery").MustString()
+	request, err := createRequest(req, query)
+	if err != nil {
+		return nil, err
+	}
+
+	response, err := ctxhttp.Do(ctx, httpClient, request)
+	if err != nil {
+		return nil, err
+	}
+
+	if response.StatusCode != http.StatusOK {
+		return nil, fmt.Errorf("invalid status code. status: %v", response.Status)
+	}
+
+	defer response.Body.Close()
```

Comment:
it must be placed before StatusCode check

Reply:
Why does this need to be placed before the statuscode check? I haven't opened a reader for the body at this point yet.

Reply:
Yikes, I will need to read body and close: https://stackoverflow.com/questions/33238518/what-could-happen-if-i-dont-close-response-body